File: app/src/main/java/com/balicodes/quicksms/ShortcutHandlerActivity.java
package com.balicodes.quicksms;

import android.app.Activity;
import android.app.AlertDialog;
import android.content.BroadcastReceiver;
import android.content.ContentUris;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.IntentFilter;
import android.net.Uri;
import android.os.Bundle;
import android.util.Log;
import android.widget.Toast;

/**
 * Created by eka on 6/25/16.
 */
public class ShortcutHandlerActivity extends Activity {
    private int currentSendingCount = 0;
    private SMSItem smsItem;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        Intent intent = getIntent();
        Uri uri = intent.getData();
        long smsID = ContentUris.parseId(uri);

        if (smsID != 0) {
            DBHelper dbHelper = new DBHelper(this);
            smsItem = dbHelper.get(smsID);
            dbHelper.close();

            AlertDialog.Builder builder = new AlertDialog.Builder(this);
            builder.setTitle(R.string.confirm_sending);

            builder.setPositiveButton(R.string.yes, new DialogInterface.OnClickListener() {
                public void onClick(DialogInterface dialog, int id) {
                    if (smsItem != null) {
                        // Sent via sending service
                        Intent intent = new Intent(ShortcutHandlerActivity.this, SendingService.class);
                        intent.putExtra(Config.SMS_BUNDLE_EXTRA_KEY, smsItem.toBundle());

                        startService(intent);
                    }
                    finish();
                }
            });
            builder.setNegativeButton(R.string.no, new DialogInterface.OnClickListener() {
                public void onClick(DialogInterface dialog, int id) {
                    finish();
                }
            });
            AlertDialog dialog = builder.create();
            dialog.show();
        } else {
            finish();
        }
    }
}

File: app/src/main/java/com/balicodes/quicksms/ShortcutHandlerActivity.kt
/*
 * Copyright (C) 2016-2018 Eka Putra
 *
 * Quick SMS is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.balicodes.quicksms

import android.app.Activity
import android.app.AlertDialog
import android.arch.lifecycle.Lifecycle
import android.arch.lifecycle.LifecycleOwner
import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.content.ContentUris
import android.content.Intent
import android.os.Bundle
import android.support.v4.app.FragmentActivity
import android.support.v7.app.AppCompatActivity
import android.util.Log
import com.balicodes.quicksms.entity.MessageEntity
import com.balicodes.quicksms.repository.MessageRepository
import com.balicodes.quicksms.service.SendingService
import com.balicodes.quicksms.viewmodel.MessageViewModel

class ShortcutHandlerActivity : FragmentActivity() {

    private lateinit var messageRepository: MessageRepository

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        messageRepository = MessageRepository(this.application)

        val uri = intent.data
        val smsID = ContentUris.parseId(uri)

        Log.d(this.localClassName.toString(), smsID.toString())

        if (smsID != 0L) {
            messageRepository.getMessage(smsID).observe(this, Observer {
                if (it != null) {
                    showConfirmSendingDialog(it)
                } else {
                    showNotFoundDialog()
                }
            })
        } else {
            finish()
        }
    }

    private fun showConfirmSendingDialog(messageEntity: MessageEntity) {

        val builder = AlertDialog.Builder(this)
        builder.setTitle(R.string.confirm_sending)

        builder.setNegativeButton(R.string.no) { _, _ -> finish() }
        builder.setOnCancelListener { finish() }

        builder.setPositiveButton(R.string.yes) { _, _ ->
            // Sent via sending service
            val intent = Intent(this, SendingService::class.java)
            intent.putExtra(Config.SMS_BUNDLE_EXTRA_KEY, messageEntity.toSmsItem().toBundle())
            this.startService(intent)

            finish()
        }

        val dialog = builder.create()
        dialog.show()
    }

    private fun showNotFoundDialog() {
        val builder = AlertDialog.Builder(this)
        builder.setTitle("Message is not available anymore.")

        builder.setOnCancelListener { finish() }

        builder.setPositiveButton(R.string.ok) { _, _ ->
            finish()
        }

        val dialog = builder.create()
        dialog.show()
    }
}
